### DISCUSSION TITLE
a methord to avoid  regenerating the whole leveldb for .idx file.

### DISCUSSION THREAD

**Discussion by guol-fnst** [seaweedfs/seaweedfs#3316] [Ideas]
In out production env,  we use -index=leveldbLarge to speeding up loading volumes. A bucket  may store billions of pictures (size:4KB-10KB). So every idx file size is very huge as large as 300MB.
When doing HA tests, we rebooting server now and then when writing or  reading object.
I found volume process  need to regenerate the whole  leveldb from index  frequently. Due to large amount of big .idx files, loading volumes costs 30 mintues  or even hours.At this time, cpu usage can reach 80-90%. This  unacceptable in production env. 
So I am trying to figure out how to  avoid  regenerating the whole leveldb for .idx file. Here is my logic. If it's unpractical  or has shortages, plesase point out.

The key word is when putting/deleting every 10000（or more ） objects, set a milestone in leveldb. 
The milestone repesents  the amount of  synced entries  between .idx and leveldb.
Next time ,when volume process needs to regenerate leveldb, we just read the milestore from leveldb, and regenerate leveldb from the offset=milestone*types.NeedleMapEntrySize of  idx file. Instead of from the very begining. 
When vacuuming volumes, leveldb files are deleted, this above logic has no influence on vacuming volumes.
This methord can reduce volume loading time from hours to  minutes ,and save lots of cpu and disk IO.

data structure
1 Add  recordNum  to  LevelDbNeedleMap struct, it equals  the number of entries  of index file. It's intialized when begain loading volumes(in func NewLevelDbNeedleMap).
2 Add a const int "milestoneCnt", which represents  the amount of entries for very miletone
```
const milestoneCnt = 100000 
type LevelDbNeedleMap struct {
	baseNeedleMapper
	dbFileName string
	db         *leveldb.DB
	recordNum  uint64
}

```

**Comment by shichanglin5** [comment]
at `weed/storage/needle_map_metric.go:123`
```go
	for remainingCount >= 0 {
		_, e := r.ReadAt(bytes[:NeedleMapEntrySize*nextBatchSize], NeedleMapEntrySize*remainingCount)
		// glog.V(0).Infoln("file", r.Name(), "readerOffset", NeedleMapEntrySize*remainingCount, "count", count, "e", e)
		if e != nil {
			return e
		}
		for i := int(nextBatchSize) - 1; i >= 0; i-- {
			key, offset, size := idx.IdxFileEntry(bytes[i*NeedleMapEntrySize : i*NeedleMapEntrySize+NeedleMapEntrySize])
			if e = fn(key, offset, size); e != nil {
				return e
			}
		}
		nextBatchSize = batchSize
		remainingCount -= nextBatchSize
	}
```
Reading the .idx file in full (or several times depending on the file size) instead of loading each `IdxFileEntry` once, should give a lot of improvement

**Reply by guol-fnst** [reply]
Thanks for the advice!
In function "generateLevelDbFile" , weeedfs use "WalkIndexFile" to traverse entries. 
https://github.com/chrislusf/seaweedfs/blob/aca20cd9f46d3eb890a7aed2921e35e85fc29dc9/weed/storage/idx/walk.go#L12
This function reads 1000 needle map entries every time, just like you mentioned.

**Comment by chrislusf** [comment]
@guol-fnst the snapshot approach seems complicated.

During volume server start, the leveldb should be re-generated only when the idx file is newer than the leveldb. See `isLevelDbFresh()` in `needle_map_leveldb.go`. I wonder why it always regenerated.

As @shichanglin5  mentioned, reverse traversing the `.idx` file is also a good idea, if this is needed.

**Reply by guol-fnst** [reply]
> See `isLevelDbFresh()` in `needle_map_leveldb.go`. I wonder why it always regenerated.

My description may be inaccuracy. Not "frequently" but sometimes, because under high pressure (5000 OPS/s), it may happen.  
Though it  doesn't happen always, but once it happend, It may  suspend the service for hours.So this is what bothers me these days.
Traversing the .idx file doesn't cost the much time. Inserting needle map entry to  leveldb costs the most time.
Reverse traversing the .idx file is my first  thought, but I found that this methord can't make sure every entry in leveldb is update to  .idx file unless traversing the full .idx file.
Besides, reverse traversing has some shortages. If a file was uploaded and  deleted right away,  reverse traversing will restore the size of the deleted file in leveldb. Acutally this file has been deleted.

**Reply by chrislusf** [reply]
when reverse traversing the idx, you can use a hashmap, or bloomfilter, to remember which key has been processed already and skip it.

**Reply by guol-fnst** [reply]
Got it! Thanks, I'll implent it later.
